File: Java/api/treehopper.libraries/src/main/java/io/treehopper/libraries/sensors/temperature/Mlx90614.java
package io.treehopper.libraries.sensors.temperature;

import io.treehopper.SMBusDevice;
import io.treehopper.interfaces.I2c;

/**
 * Melexis MLX90614 non-contact I2c temperature sensor
 */
public class Mlx90614 {

    public class TempRegister extends TemperatureSensor
    {
        private SMBusDevice dev;
        private byte register;

        public TempRegister(SMBusDevice dev, byte register) {
            this.dev = dev;
            this.register = register;
        }

        @Override
        public void update() {
            int data = dev.readWordData(register);
            data &= 0x7FFF; // chop off the error bit of the high byte
            this.celsius = data * 0.02 - 273.15;
        }
    }

    SMBusDevice dev;
    private Temperature ambient;
    private Temperature object;

    public Mlx90614(I2c module) {
        this.dev = new SMBusDevice((byte)0x5A, module, 30);

        ambient = new TempRegister(this.dev, (byte)0x06);
        object = new TempRegister(this.dev, (byte)0x07);
    }

	public Temperature getAmbient() {
		return ambient;
	}

	public Temperature getObject() {
		return object;
	}

}

File: C++/API/src/libraries/Sensors/Temperature/Mlx90614.cpp
#include "Libraries/Sensors/Temperature/Mlx90614.h"

namespace Treehopper {
    namespace Libraries {
        namespace Sensors {
            namespace Temperature {
                Mlx90614::Mlx90614(I2c &i2c, int rate) :
                        dev(0x5B, i2c, rate),
                        ambient(dev, 0x06),
                        object(dev, 0x07) {}


                Mlx90614::TempRegister::TempRegister(SMBusDevice &dev, uint8_t reg)
                        : dev(dev),
                          reg(reg) {}

                void Mlx90614::TempRegister::update() {
                    auto data = dev.readWordData(reg);
                    data &= 0x7FFF; // chop off the error bit of the high byte
                    _celsius = data * 0.02 - 273.15;
                }
            }
        }
    }
}
